### ISSUE TITLE
Macro expansion error inside #[async_trait] item points to wrong span

### ISSUE THREAD

**Issue by tmandry** [dtolnay/async-trait#65]
The following code:

```rust
use async_trait::async_trait;

macro_rules! picky {
    { stuff: { $contents:expr } } => { println!("stuff: {}", $contents); }
}

#[async_trait]
trait Foo {
    async fn foo() -> i32;
}

struct Bar;

#[async_trait]
impl Foo for Bar {
    async fn foo() -> i32 {
        picky!({ 123 });
        42
    }
}
```

emits the following error:

```
error: no rules expected the token `{`
  --> src/main.rs:16:1
   |
5  | macro_rules! picky {
   | ------------------ when calling this macro
...
16 | #[async_trait]
   | ^^^^^^^^^^^^^^ no rules expected this token in macro call
```

Can you spot the problem? We forgot `stuff:` inside `picky!()`. However, the "no rules expected this token" note points to `#[async_trait]` instead of the unexpected `{` token.

When I make the async fn's non-async (leaving `#[async_trait]`), the error message looks correct. I also can't reproduce using normal macros, so I suspect the issue is with async-trait.

**Comment by dtolnay** [comment]
Thanks for the report. Looks like an async-trait or syn bug. Based on putting:

```diff
 #[proc_macro_attribute]
 pub fn async_trait(args: TokenStream, input: TokenStream) -> TokenStream {
+    eprintln!("{:#?}", input);
     let args = parse_macro_input!(args as Args);
     let mut item = parse_macro_input!(input as Item);
     expand(&mut item, args.local);
+    eprintln!("{:#?}", quote!(#item));
     TokenStream::from(quote!(#item))
 }
```

in the proc macro, I can see `picky!({ 123 })` coming in as:

```rust
Ident {
    ident: "picky",
    span: #0 bytes(270..275),
},
Punct {
    ch: '!',
    spacing: Alone,
    span: #0 bytes(275..276),
},
Group {
    delimiter: Parenthesis,
    stream: TokenStream [
        Group {
            delimiter: Brace,
            stream: TokenStream [
                Literal {
                    lit: Lit {
                        kind: Integer,
                        symbol: "123",
                        suffix: None,
                    },
                    span: #0 bytes(279..282),
                },
            ],
            span: #0 bytes(277..284),
        },
    ],
    span: #0 bytes(276..285),
},
```

and going back out as:

```rust
Ident {
    ident: "picky",
    span: #0 bytes(270..275),
},
Punct {
    ch: '!',
    spacing: Alone,
    span: #0 bytes(275..276),
},
Group {
    delimiter: Parenthesis,
    stream: TokenStream [
        Group {
            delimiter: Brace,
            stream: TokenStream [
                Literal {
                    lit: Lit {
                        kind: Integer,
                        symbol: "123",
                        suffix: None,
                    },
                    span: #0 bytes(279..282),
                },
            ],
            span: #7 bytes(200..214),
        },
    ],
    span: #0 bytes(276..285),
},
```

Notice the wrong span `#7 bytes(200..214)` on the Brace token.

**Comment by dtolnay** [comment]
Fixed in async-trait 0.1.23.

**Comment by tmandry** [comment]
Thanks for fixing this so quickly!
